File: apps/ui/modules/happier-hosted-web-frame/ios/HappierHostedWebFrameModule.swift
import ExpoModulesCore

/**
 Expo entrypoint for the one token-scoped hosted-web Artifact frame. The
 registry remains the native consumer of opaque Artifact cache coordinates;
 this module deliberately has no URL, byte, or cache-path API.
 */
public final class HappierHostedWebFrameModule: Module {
  public func definition() -> ModuleDefinition {
    Name("HappierHostedWebFrame")

    // The strict typed admission result, identical in shape to Android/JS. iOS
    // has no profile-isolation capability class, so it reports only the two
    // outcomes its registry can produce.
    AsyncFunction("registerArtifact") { (input: [String: Any]) -> [String: Any] in
      if HostedWebArtifactRegistryOwner.shared.register(input) {
        return ["kind": "registered"]
      }
      return [
        "kind": "unavailable",
        "code": "native_artifact_resource_registration_failed",
      ]
    }

    // This is intentionally a synchronous acknowledgement. Artifact keeps its
    // handle/cache index until the native registry rejects later reads.
    Function("unregisterArtifact") { (token: String) -> Bool in
      HostedWebArtifactRegistryOwner.shared.unregister(token)
    }

    View(HostedWebArtifactView.self) {
      Events("onMessage", "onLoadStart", "onLoadEnd", "onLoadError", "onExternalNavigation", "onBlockedNavigation", "onHistoryStateChange")

      Prop("title") { (view: HostedWebArtifactView, title: String?) in
        view.setTitle(title)
      }
      Prop("artifactHandleToken") { (view: HostedWebArtifactView, token: String?) in
        view.setArtifactHandleToken(token)
      }
      Prop("initialPathAndQuery") { (view: HostedWebArtifactView, pathAndQuery: String?) in
        view.setInitialPathAndQuery(pathAndQuery)
      }
      Prop("allowedNavigationOrigins") { (view: HostedWebArtifactView, origins: [String]) in
        view.setAllowedNavigationOrigins(origins)
      }

      // Expo view AsyncFunctions live on the view ref and execute on the main
      // queue. The JS component invokes this only while its host bridge is
      // mounted; the view independently verifies its token and active origin.
      AsyncFunction("postHostMessage") { (view: HostedWebArtifactView, serializedMessage: String) -> Bool in
        view.postHostMessage(serializedMessage)
      }
      AsyncFunction("goBack") { (view: HostedWebArtifactView) -> Bool in
        view.goBack()
      }
    }

    OnDestroy {
      HostedWebArtifactRegistryOwner.shared.clear()
    }
  }
}

File: apps/ui/modules/happier-hosted-web-frame/android/src/main/java/dev/happier/hostedwebframe/HappierHostedWebFrameModule.kt
package dev.happier.hostedwebframe

import android.os.Handler
import android.os.Looper
import expo.modules.kotlin.functions.Queues
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

/**
 * The only JavaScript-to-native registration seam for hosted-web Artifact
 * frames. Registration receives opaque Artifact metadata only; the exported
 * view resolves one registered token through [HostedWebArtifactRegistryOwner].
 */
class HappierHostedWebFrameModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("HappierHostedWebFrame")

    OnCreate {
      // ProfileStore is UI-thread-only. A bounded prefix sweep repairs only
      // frames orphaned by a prior process death before this module can mount.
      val cleanup = Runnable { HostedWebArtifactView.cleanupOrphanedProfiles() }
      if (Looper.myLooper() == Looper.getMainLooper()) {
        cleanup.run()
      } else {
        Handler(Looper.getMainLooper()).post(cleanup)
      }
    }

    AsyncFunction("registerArtifact") { input: Map<String, Any?> ->
      val profileIsolationCapability = HostedWebArtifactView.profileIsolationUnavailableCapability()
      if (profileIsolationCapability != null) {
        return@AsyncFunction mapOf(
          "kind" to "unavailable",
          "code" to "hosted_web_profile_isolation_unavailable",
          "capability" to profileIsolationCapability
        )
      }
      val context = appContext.reactContext ?: return@AsyncFunction mapOf(
        "kind" to "unavailable",
        "code" to "native_artifact_resource_registration_failed"
      )
      if (HostedWebArtifactRegistryOwner.register(context, input)) {
        mapOf("kind" to "registered")
      } else {
        mapOf(
          "kind" to "unavailable",
          "code" to "native_artifact_resource_registration_failed"
        )
      }
    }.runOnQueue(Queues.DEFAULT)

    // This acknowledgement is intentionally synchronous. Artifact removes its
    // own token/index only after true, which means a later native read cannot
    // observe the token even if storage cleanup follows afterwards.
    Function("unregisterArtifact") { token: String ->
      val context = appContext.reactContext ?: return@Function false
      HostedWebArtifactRegistryOwner.unregister(context, token)
    }

    View(HostedWebArtifactView::class) {
      Events("onMessage", "onLoadStart", "onLoadEnd", "onLoadError", "onExternalNavigation", "onBlockedNavigation", "onHistoryStateChange")

      Prop<String?>("title") { view, title ->
        view.setTitle(title)
      }
      Prop<String?>("artifactHandleToken") { view, token ->
        view.setArtifactHandleToken(token)
      }
      Prop<String?>("initialPathAndQuery") { view, pathAndQuery ->
        view.setInitialPathAndQuery(pathAndQuery)
      }
      Prop<List<String>>("allowedNavigationOrigins") { view, origins ->
        view.setAllowedNavigationOrigins(origins)
      }

      AsyncFunction("postHostMessage") { view: HostedWebArtifactView, serializedMessage: String ->
        view.postHostMessage(serializedMessage)
      }
      AsyncFunction("goBack") { view: HostedWebArtifactView ->
        view.goBack()
      }

      OnViewDestroys { view ->
        view.dispose()
      }
    }

    OnDestroy {
      HostedWebArtifactRegistryOwner.clear()
    }
  }
}
